File: fsharp/FsSearch/FsSearch/SearchResultFormatter.fs
namespace FsSearch

type SearchResultFormatter (settings : SearchSettings.t) =

    let Colorize (s : string) (matchStartIndex : int) (matchEndIndex : int) : string =
        let matchLength = matchEndIndex - matchStartIndex
        s.Substring(0, matchStartIndex) +
            Color.Green + 
            s.Substring(matchStartIndex, matchLength) +
            Color.Reset + 
            s.Substring(matchStartIndex + matchLength)

    let GetRelativeFilePath (result : SearchResult.t) : string =
        if settings.StartPath <> null then
            if settings.StartPath = "~"
            then FileUtil.ContractPath result.File.File.FullName
            else FileUtil.GetRelativePath result.File.File.FullName settings.StartPath
        else result.File.File.FullName
                
    
    let FormatMatchingLine (result : SearchResult.t) : string =
        let mutable formatted : string = result.Line.Trim()
        let mutable formattedLength = formatted.Length
        let leadingWhitespaceCount = result.Line.TrimEnd().Length - formattedLength
        let maxLineEndIndex = formattedLength - 1
        let matchLength = result.MatchEndIndex - result.MatchStartIndex
        let matchStartIndex = result.MatchStartIndex - 1 - leadingWhitespaceCount
        let matchEndIndex = matchStartIndex + matchLength
        let rec recGetIndices (lineStartIndex : int) (lineEndIndex : int) (matchStartIndex : int) (matchEndIndex : int) : (int * int * int * int) =
            if lineEndIndex - lineStartIndex < settings.MaxLineLength then
                let lsi =
                    if lineStartIndex > 0
                    then lineStartIndex - 1
                    else lineStartIndex
                let lei =
                    if lineEndIndex - lsi < settings.MaxLineLength && lineEndIndex < maxLineEndIndex
                    then lineEndIndex + 1
                    else lineEndIndex
                let (msi, mei) =
                    if lineStartIndex > 0
                    then (matchStartIndex + 1, matchEndIndex + 1)
                    else (matchStartIndex, matchEndIndex)
                recGetIndices lsi lei msi mei
            else
                (lineStartIndex, lineEndIndex, matchStartIndex, matchEndIndex)
        let (lineStartIndex, lineEndIndex, matchStartIndex, matchEndIndex) =
            if formattedLength > settings.MaxLineLength then
                recGetIndices matchStartIndex matchEndIndex matchStartIndex matchEndIndex
            else
                (0, maxLineEndIndex, matchStartIndex, matchEndIndex)
        formattedLength <- if lineStartIndex = 0 then lineEndIndex - lineStartIndex + 1 else lineEndIndex - lineStartIndex
        formatted <- formatted.Substring(lineStartIndex, formattedLength)
        if lineStartIndex > 2 then
            formatted <- "..." + formatted.Substring(3)
        if lineEndIndex < maxLineEndIndex - 3 then
            formatted <- formatted.Substring(0, formattedLength - 3) + "..."
        if settings.Colorize then
            formatted <- Colorize formatted matchStartIndex matchEndIndex
        formatted

    let SingleLineFormat (result : SearchResult.t) : string =
        let matchString =
            if result.LineNum = 0 then
                sprintf " matches at [%d:%d]" result.MatchStartIndex result.MatchEndIndex
            else
                let line = FormatMatchingLine result
                sprintf ": %d: [%d:%d]: %s" result.LineNum result.MatchStartIndex result.MatchEndIndex line
        (GetRelativeFilePath result) + matchString

    let MultLineFormat (result : SearchResult.t) : string =
        let hdr = 
            String.concat "\n" [
                sprintf "%s" (new string('=', 80));
                sprintf "%s: %d: [%d:%d]" (GetRelativeFilePath result) result.LineNum result.MatchStartIndex result.MatchEndIndex;
                sprintf "%s" (new string('-', 80));
            ] + "\n"
        let maxLineNum = result.LineNum + (List.length result.LinesAfter)
        let maxLineNumLength = maxLineNum.ToString().Length
        let paddedLineNum (lineNum : int) : string =
            let lineNumString = lineNum.ToString()
            (new string(' ', (maxLineNumLength - lineNumString.Length))) + lineNumString
        let rec recLines (lines : string list) (lineNum : int) (linesString : string) : string =
            match lines with
            | [] -> linesString
            | l :: ls ->
                recLines ls (lineNum + 1) (linesString + (sprintf "  %s | %s\n" (paddedLineNum lineNum) l))
        let linesBeforeString = recLines result.LinesBefore (result.LineNum - (List.length result.LinesBefore)) ""
        let linesAfterString = recLines result.LinesAfter (result.LineNum + 1) ""
        String.concat "" [
            hdr;
            linesBeforeString;
            sprintf "> %s | %s\n" (paddedLineNum result.LineNum) result.Line;
            linesAfterString;
        ]

    member this.Format (result : SearchResult.t) : string =
        if (not (List.isEmpty result.LinesBefore)) || (not (List.isEmpty result.LinesAfter)) then
            MultLineFormat result
        else
            SingleLineFormat result
;;

File: objc/objcsearch/objcsearch/SearchResultFormatter.m
#import "FileUtil.h"
#import "SearchResultFormatter.h"

@implementation SearchResultFormatter

- (instancetype) initWithSettings:(SearchSettings*)settings
{
    self = [super init];
    if (self) {
        self.settings = settings;
    }
    return self;
}

- (NSString *) format:(SearchResult*)result {
    if (result.linesBefore.count > 0 || result.linesAfter.count > 0) {
        return [self multiLineFormat:result];
    } else {
        return [self singleLineFormat:result];
    }
}

- (NSString *) getRepeatingString:(NSString*)s count:(int)count {
    return [@"" stringByPaddingToLength:count withString:s startingAtIndex:0];
}

- (NSString *) singleLineFormat:(SearchResult*)result {
    NSMutableString *s = [NSMutableString string];
    [s appendString:[result getFilePath]];
    if (result.lineNum > 0) {
        [s appendFormat:@": %lu: [%lu:%lu]: ", result.lineNum,
         result.matchStartIndex, result.matchEndIndex];
//        [s appendString:[self.line stringByTrimmingCharactersInSet:[NSCharacterSet characterSetWithCharactersInString:@" \t\r\n"]]];
        [s appendString:[self formatMatchingLine:result]];
    } else {
        [s appendFormat:@" matches at [%lu:%lu]",
         result.matchStartIndex, result.matchEndIndex];
    }
    return [NSString stringWithString:s];
}

- (NSString *) formatMatchingLine:(SearchResult*)result {
    NSCharacterSet *whitespaceCharacterSet = [NSCharacterSet characterSetWithCharactersInString:@" \t\r\n"];
    NSMutableString *formatted = [NSMutableString string];
    int leadingWhitespaceCount = 0;
    while ([whitespaceCharacterSet characterIsMember:[result.line characterAtIndex:leadingWhitespaceCount]]) {
        leadingWhitespaceCount++;
    }
    [formatted appendString:[result.line stringByTrimmingCharactersInSet:whitespaceCharacterSet]];
    long formattedLength = [formatted length];
    long maxLineEndIndex = formattedLength - 1;
    long matchLength = result.matchEndIndex - result.matchStartIndex;
    long matchStartIndex = result.matchStartIndex - 1 - leadingWhitespaceCount;
    long matchEndIndex = matchStartIndex + matchLength;

    if (formattedLength > self.settings.maxLineLength) {
        long lineStartIndex = matchStartIndex;
        long lineEndIndex = lineStartIndex + matchLength;
        matchStartIndex = 0;
        matchEndIndex = matchLength;

        while (lineEndIndex > formattedLength - 1) {
            lineStartIndex--;
            lineEndIndex--;
            matchStartIndex++;
            matchEndIndex++;
        }

        formattedLength = lineEndIndex - lineStartIndex;
        while (formattedLength < self.settings.maxLineLength) {
            if (lineStartIndex > 0) {
                lineStartIndex--;
                matchStartIndex++;
                matchEndIndex++;
                formattedLength = lineEndIndex - lineStartIndex;
            }
            if (formattedLength < self.settings.maxLineLength && lineEndIndex < maxLineEndIndex) {
                lineEndIndex++;
            }
            formattedLength = lineEndIndex - lineStartIndex;
        }

        formatted = [NSMutableString stringWithFormat:@"%@", [formatted substringWithRange:NSMakeRange(lineStartIndex, formattedLength)]];

        if (lineStartIndex > 2) {
            formatted = [NSMutableString stringWithFormat:@"...%@", [formatted substringFromIndex:3]];
        }
        if (lineEndIndex < maxLineEndIndex - 3) {
            formatted = [NSMutableString stringWithFormat:@"%@...", [formatted substringToIndex:(formattedLength - 3)]];
        }
    }

    if (self.settings.colorize) {
        NSMutableString *c = [NSMutableString stringWithFormat:@"%@", [formatted substringToIndex:matchStartIndex]];
//        [c appendString:ANSI_GREEN];
        [c appendFormat:@"%s", ANSI_GREEN];
        [c appendString:[formatted substringWithRange:NSMakeRange(matchStartIndex, matchLength)]];
        [c appendString:ANSI_RESET];
        [c appendString:[formatted substringFromIndex:matchEndIndex]];
        formatted = [NSMutableString stringWithFormat:@"%@", c];
    }

    return [NSString stringWithString:formatted];
}

- (NSString *) multiLineFormat:(SearchResult*)result {
    NSMutableString *s = [NSMutableString string];
    int sepLen = 80;
    [s appendFormat:@"%@\n", [self getRepeatingString:@"=" count:sepLen]];
    [s appendString:[result getFilePath]];
    [s appendFormat:@": %lu: [%lu:%lu]\n", result.lineNum,
     result.matchStartIndex, result.matchEndIndex];
    [s appendFormat:@"%@\n", [self getRepeatingString:@"-" count:sepLen]];
    
    long maxLineNum = result.lineNum + [result.linesAfter count];
    long maxNumLen = [[NSString stringWithFormat:@"%lu", maxLineNum] length];
    char cNumFormat[maxNumLen + 4];
    sprintf(cNumFormat, "%%%lulu", maxNumLen);
    NSString *numberFormat = [NSString stringWithUTF8String:cNumFormat];
    for (unsigned long i=0; i < [result.linesBefore count]; i++) {
        [s appendString:[self formatLineString:result.linesBefore[i]
                                       lineNum:result.lineNum - [result.linesBefore count] + i
                                  numberFormat:numberFormat prefix:@"  "]];
    }
    [s appendString:[self formatLineString:result.line
                                   lineNum:result.lineNum
                              numberFormat:numberFormat prefix:@"> "]];
    for (unsigned long i=0; i < [result.linesAfter count]; i++) {
        [s appendString:[self formatLineString:result.linesAfter[i]
                                       lineNum:result.lineNum + i + 1
                                  numberFormat:numberFormat prefix:@"  "]];
    }

    return [NSString stringWithString:s];
}

- (NSString *) formatLineString:(NSString*)line
                        lineNum:(long)lineNum
                   numberFormat:(NSString*)numberFormat
                         prefix:(NSString*)prefix {
    NSMutableString *s = [NSMutableString stringWithString:prefix];
    char cNumString[10];
    sprintf(cNumString, [numberFormat UTF8String], lineNum);
    NSString *numberString = [NSString stringWithUTF8String:cNumString];
    [s appendFormat:@"%@ | %@", numberString, line];
    return [NSString stringWithString:s];
}

@end
